crates/cubecl-wgpu/src/compiler/wgsl/compiler.rs

```diff
@@ -720,267 +732,310 @@ impl WgslCompiler {
                 lhs: self.compile_variable(op.lhs),
                 rhs: self.compile_variable(op.rhs),
                 out: self.compile_variable(out),
-            },
-            cube::Operator::GreaterEqual(op) => wgsl::Instruction::GreaterEqual {
+            }),
+            cube::Operator::LowerEqual(op) => instructions.push(wgsl::Instruction::LowerEqual {
                 lhs: self.compile_variable(op.lhs),
                 rhs: self.compile_variable(op.rhs),
                 out: self.compile_variable(out),
-            },
-            cube::Operator::NotEqual(op) => wgsl::Instruction::NotEqual {
+            }),
+            cube::Operator::GreaterEqual(op) => {
+                instructions.push(wgsl::Instruction::GreaterEqual {
+                    lhs: self.compile_variable(op.lhs),
+                    rhs: self.compile_variable(op.rhs),
+                    out: self.compile_variable(out),
+                })
+            }
+            cube::Operator::NotEqual(op) => instructions.push(wgsl::Instruction::NotEqual {
                 lhs: self.compile_variable(op.lhs),
                 rhs: self.compile_variable(op.rhs),
                 out: self.compile_variable(out),
-            },
-            cube::Operator::Cast(op) => wgsl::Instruction::Assign {
+            }),
+            cube::Operator::Cast(op) => instructions.push(wgsl::Instruction::Assign {
                 input: self.compile_variable(op.input),
                 out: self.compile_variable(out),
-            },
-            cube::Operator::IndexAssign(op) => wgsl::Instruction::IndexAssign {
-                lhs: self.compile_variable(op.lhs),
-                rhs: self.compile_variable(op.rhs),
-                out: self.compile_variable(out),
-            },
-            cube::Operator::UncheckedIndexAssign(op) => wgsl::Instruction::IndexAssign {
+            }),
+            cube::Operator::Index(op) => {
+                if matches!(self.strategy, ExecutionMode::Checked) && op.lhs.has_length() {
+                    let lhs = op.lhs;
+                    let rhs = op.rhs;
+                    let array_len = scope.create_local(cube::Item::new(u32::as_elem()));
+
+                    instructions.extend(self.compile_scope(scope));
+
+                    let length = match lhs.has_buffer_length() {
+                        true => cube::Metadata::BufferLength { var: lhs },
+                        false => cube::Metadata::Length { var: lhs },
+                    };
+
+                    instructions.push(self.compile_metadata(length, Some(array_len)));
+                    instructions.push(wgsl::Instruction::CheckedIndex {
+                        len: self.compile_variable(array_len),
+                        lhs: self.compile_variable(lhs),
+                        rhs: self.compile_variable(rhs),
+                        out: self.compile_variable(out),
+                    });
```

Comment:
I don't think it's necessary right now, but I think having this implemented with a `cpa` kernel like `CheckedIndexAssign` would be easier.

Reply:
I'm also not sure when we don't have `BufferLength`

Reply:
+1 to cpa someday.

And yeah this behaviour is copied from the CPP version. Personally I feel like that kind of thing should fail to compile, rather than creating a language with strange edge cases of UB. I'm not sure what kind of situation means you don't have buffer length